controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/DeleteReaderGroupTask.java

```diff
@@ -0,0 +1,79 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.eventProcessor.requesthandlers;
+
+import com.google.common.base.Preconditions;
+
+import io.pravega.client.stream.ReaderGroupConfig;
+import io.pravega.client.stream.Stream;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.controller.store.stream.RGOperationContext;
+import io.pravega.controller.store.stream.StreamMetadataStore;
+import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.shared.NameUtils;
+import io.pravega.shared.controller.event.DeleteReaderGroupEvent;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Iterator;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Request handler for executing a delete operation for a ReaderGroup.
+ */
+@Slf4j
+public class DeleteReaderGroupTask implements ReaderGroupTask<DeleteReaderGroupEvent> {
+
+    private final StreamMetadataStore streamMetadataStore;
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final ScheduledExecutorService executor;
+
+    public DeleteReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
+                                 final StreamMetadataStore streamMetaStore,
+                                 final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetaStore);
+        Preconditions.checkNotNull(streamMetaTasks);
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataStore = streamMetaStore;
+        this.streamMetadataTasks = streamMetaTasks;
+        this.executor = executor;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(final DeleteReaderGroupEvent request) {
+      String scope = request.getScope();
+      String readerGroup = request.getRgName();
+      long requestId = request.getRequestId();
+      log.info("Inside Delete Event");
+      final RGOperationContext context = streamMetadataStore.createRGContext(scope, readerGroup);
+      return streamMetadataStore.getReaderGroupConfigRecord(scope, readerGroup, context, executor)
+               .thenCompose(configRecord -> {
+               if (!ReaderGroupConfig.StreamDataRetention.values()[configRecord.getObject().getRetentionTypeOrdinal()]
+                       .equals(ReaderGroupConfig.StreamDataRetention.NONE)) {
+               String scopedRGName = NameUtils.getScopedReaderGroupName(scope, readerGroup);
+               log.info("Inside Delete Event, found config");
+               // update Stream metadata tables, if RG is a Subscriber
+               Iterator<String> streamIter = configRecord.getObject().getStartingStreamCuts().keySet().iterator();
+               return Futures.loop(() -> streamIter.hasNext(), () -> {
+                       Stream stream = Stream.of(streamIter.next());
+                       return streamMetadataStore.deleteSubscriber(stream.getScope(),
```

Comment:
again the delete subscriber should pass the generation number so that unsubscription is also conditional and we do not unnecessarily remove a subscription unintendedly.

Reply:
fixed